BinaryEncoding.md

```diff
@@ -436,34 +436,62 @@ The expectation is that, when a binary WebAssembly module is viewed in a browser
 environment, the data in this section will be used as the names of functions
 and locals in the [text format](TextFormat.md).
 
+The name section contains a sequence of name subsections:
+
 | Field | Type | Description |
 | ----- | ---- | ----------- |
-| count | `varuint32` | count of entries to follow |
-| entries | `function_names*` | sequence of names |
+| name_type | `varuint7` | code identifying type of name contained in this subsection |
+| name_payload_len | `varuint32` | size of this subsection in bytes |
+| name_payload_data | `bytes` | content of this section, of length `name_payload_len` |
+
+Since name subsections have a given length, unknown or unwanted names can be
+skipped over by an engine. The current list of valid `name_type` codes are:
+
+| Name Type | Code | Description |
+| --------- | ---- | ----------- |
+| [Function](#function-names) | `1` | Assigns names to functions |
+| [Local](#local-names) | `2` | Assigns names to locals in functions |
+
+When present, name subsections must appear in this order and at most once. The
+end of the last subsection must coincide with the last byte of the name
+section to be a well-formed name section.
+
+#### Name List
+
+In any of the subsequent subsections, a `name_list` is encoded as:
 
-The sequence of `function_names` assigns names to the corresponding
-[function index](Modules.md#function-index-space). (Note: this assigns names to both
-imported and module-defined functions.) The count may differ from the actual number of functions.
+| Name Type | Type | Description |
+| --------- | ---- | ----------- |
+| count | `varuint32` | number of `name` in names |
+| names | `name*` | sequence of `name` |
+
+where a `name` is encoded as:
+
+| Name Type | Type | Description |
+| --------- | ---- | ----------- |
+| name_len | `varuint32` | number of bytes in name_str |
+| name_str | `bytes` | UTF8 encoding of the name |
```

Comment:
IIUC what we had before was a hacked-up names section that got us limping along, with no real intent of keeping it around. Maybe this would help me understand better: what's the intent of this refactoring? I thought you were going for something that'll stand some test of time, so without designing DWARF / TROLL I'm trying to see what we can minimally get "right". Is that out of scope? Happy to back away if so.